firmware/src/imu/mod.rs

```diff
@@ -29,32 +32,69 @@ pub trait Imu {
 	fn quat(&mut self) -> nb::Result<Quat, Self::Error>;
 }
 
-/// Gets data from the IMU
+#[task]
 pub async fn imu_task(
-	quat_signal: &Unreliable<Quat>,
-	i2c: impl crate::aliases::I2c,
-	mut delay: impl DelayMs<u32>,
-) {
-	debug!("Started sensor_task");
-	let mut imu = ඞ::new_imu(i2c, &mut delay);
+	packets: &'static Packets,
+	i2c: crate::aliases::ඞ::I2cConcrete<'static>,
+	mut delay: crate::aliases::ඞ::DelayConcrete,
+) -> ! {
+	debug!("imu_task!");
+	let main_imu = ඞ::new_imu(i2c, &mut delay);
 	info!("Initialized IMU!");
 
+	imu_loop(0, main_imu, packets).await
+}
+
+/// Run a single IMU indefinitely
+async fn imu_loop<T: Imu>(sensor_id: u8, mut imu: T, packets: &'static Packets) -> ! {
+	// We update hosts view of sensors when this changes
+	let mut state = (false, *packets.connection.lock().await);
+	// The throttle will ensure our IMU data is only sent at 100hz, instead of as fast as possible
+	let mut throttle = embassy_time::Ticker::every(Duration::from_millis(10));
+
 	loop {
-		let q = match nb2a(|| imu.quat()).await {
-			Ok(q) => q,
-			Err(err) => {
-				warn!("Error in IMU: {}", defmt::Debug2Format(&err));
-				continue;
-			}
-		};
-		trace!(
-			"Quat values: x: {}, y: {}, z: {}, w: {}",
-			q.coords.x,
-			q.coords.y,
-			q.coords.z,
-			q.coords.w
-		);
-		quat_signal.signal(q);
-		yield_now().await // Yield to ensure fairness
+		// Wait for next IMU update
+		throttle.next().await;
+
+		// Query IMU data
+		let quat = nb2a(|| imu.quat()).await;
+
+		// Update sensor status when the IMU returned quat changes or connection is updated
+		let new_state = (quat.is_ok(), *packets.connection.lock().await);
+		if state != new_state {
+			state = new_state;
+
+			// Move to firmware_protocol?
+			let sensor_status = if quat.is_ok() {
+				SensorStatus::Ok
+			} else {
+				SensorStatus::Offline
+			};
+
+			// Let the server know
+			packets
+				.serverbound
+				.send(SbPacket::SensorInfo {
+					sensor_id,
+					sensor_status,
+					sensor_type: T::IMU_TYPE,
+				})
+				.await;
+		}
+
+		// Got valid IMU pose, send it to server
+		if let Ok(unit) = quat {
+			let quat = unit.into_inner().into();
+
+			packets
+				.serverbound
+				.send(SbPacket::RotationData {
```

Comment:
We should be doing packet stuff in the protocol task, because I want the rest of the codebase to not care at all about either the protocol or the networking.

On example why it may matter: `net-esb` for example may use a separate, simpler protocol. Or the protocol may itself be a feature flag (legacy slimevr protocol vs a new flatbuffer protocol, for example)